### ISSUE TITLE
Dynamic executor properties

### ISSUE THREAD

**Issue by AerialMantis** [executors/executors#244]
What is the current design regarding whether or not the `prefer` customisation point should alter the interface provided by the executor and are there any guarantees about whether `prefer` will return an explicit executor type or a polymorphic executor?

A use case which we are particularly interested in is that which requires the ability to dynamically request properties of an executor at runtime. This is important for generic applications which may not know the capabilities of the platform that it will be targeting at the point of compilation and also for applications which are cross compiled and are not capable of knowing platform capabilities at the point of compilation.

For this we feel it's important to be able to query an executor for support for particular properties and request for them via `prefer` and have `prefer` return a polymorphic executor.

Though where this gets complicated is when properties involve altering the interface of the executor, for which you cannot write code generically without the polymorphic executor requiring all possible execution functions, even if invalid functions simply throw an exception when called. There fore it may not make sense for this capability to be possible for properties that change the interface, so perhaps the executor returned could be a particular executor category.

**Comment by jaredhoberock** [comment]
There are two requests here and I think one of these is already possible with the functionality in the current paper.

Any "static" executor can be stored in the polymorphic `executor` wrapper. Because `executor` itself fulfills the general executor requirements, it can be used with `execution::require` & `execution::prefer`. The result of passing such an `executor` to either `execution::require` & `execution::prefer` will itself be another instance of `executor`.

The part of your request we don't have is this:

> For this we feel it's important to be able to query an executor for support for particular properties

I don't believe we have yet a general mechanism to query executors for their properties.

**Comment by AerialMantis** [comment]
As discussed in the last call i've come up with some examples of dynamic properties that we may wish to include or at least support. I've tried to focus on properties that would be applicable to a large number of executor implementations:

```
constexpr struct preferred_bulk_execution_shape_t{} preferred_bulk_execution_shape;
constexpr struct max_bulk_execution_shape_t {} max_bulk_execution_shape;
constexpr struct max_shared_parameter_size_t {} max_shared_parameter_size;
```

* `preferred_bulk_execution_shape` would return the preferred shape that users should use when calling bulk execution functions with this executor.
* `max_bulk_execution_shape` would return the maximum shape that this executor can support.
* `max_shared_parameter_size` would return the maximum size of shared type returned by the shared factory that is supported by this executor.

It's important that these properties be dynamic because these values can change greatly depending on the execution resource being executed on, even within the same vendor, so it can be important for users to query these in order to gain the best performance out of executors in a generic way.

There are many other properties that an executor implementation may want to support such as device type, vendor name or support for extended feature sets, which a dynamic property or query interface would support.

In the last call we discussed the idea of extending the existing `prefer` member function to allow it to return a property for certain properties, so I had a go at prototyping this to see how it would look:

```
#include <type_traits>

class shape {};
class size {};

constexpr struct never_blocking_t {} never_blocking;
constexpr struct possibly_blocking_t {} possibly_blocking;
constexpr struct always_blocking_t {} always_blocking;
constexpr struct preferred_bulk_execution_shape_t { shape _shape; shape get() { return _shape; } } preferred_bulk_execution_shape;
constexpr struct max_bulk_execution_shape_t { shape _shape; shape get() { return _shape; } } max_bulk_execution_shape;
constexpr struct max_shared_parameter_size_t { size _size; size get() { return _size; } } max_shared_parameter_size;

enum class property_type : unsigned int { setter, getter };

template <typename T>
constexpr property_type property_type_v = property_type::setter;

template <>
constexpr property_type property_type_v<preferred_bulk_execution_shape_t> = property_type::getter;

template <>
constexpr property_type property_type_v<max_bulk_execution_shape_t> = property_type::getter;

template <>
constexpr property_type property_type_v<max_shared_parameter_size_t> = property_type::getter;

template <typename T, property_type P>
struct prefer_member_return;

template <typename T>
struct prefer_member_return<T, property_type::setter> {
  using type = void;
};

template <typename T>
struct prefer_member_return<T, property_type::getter> {
  using type = T;
};

template <typename T>
using prefer_member_return_t = typename prefer_member_return<T, property_type_v<T>>::type;

class executor {
public:
  executor() {}

  template <typename T, typename = std::enable_if_t<property_type_v<T> == property_type::setter>>
  void prefer(T prop) {
    process_property(prop); // modify semantics of executor
    return;
  }

  template <typename T, typename = std::enable_if_t<property_type_v<T> == property_type::getter>>
  prefer_member_return_t<T> prefer(T prop) {
    process_property(prop);  // retrieve value and store it in property
    return prop;
  }
};

int main() {

  executor exec;

  exec.prefer(never_blocking);

  auto preferedShape = exec.prefer(preferred_bulk_execution_shape).get();
  
}
```

This works by having some properties marked as getters and some as setters, then using enable_if to apply the property as normal in the case of a setter property, and to retrieve the requested value, store it in the property and return the property in the case of a getter property. This seams to work very nicely as it allows the regular properties to remain unaffected and doesn't require a significant change to the interface. The only suggestion I would make is we may want to choose a different name for `prefer` if it were to have this dual functionality?

**Comment by AerialMantis** [comment]
This interface could also be used to provide a way of querying an executor for whether a preference has taken effect or not (https://github.com/executors/issaquah_2016/issues/242), by adding an additional getter property.

For example:

```
constexpr struct never_blocking_t {} never_blocking;  // setter property
constexpr struct possibly_blocking_t {} possibly_blocking;  // setter property
constexpr struct always_blocking_t {} always_blocking;  // setter property
constexpr struct never_blocking_t {} blocking;  // getter property
```
Then users could use one of the setter properties to request a particular blocking semantics and then use the getter property to query it's effect:

```
int main() {
  executor exec;

  exec.prefer(never_blocking);
  auto blockingSemantics = exec.prefer(blocking).get();
}
```

**Comment by AerialMantis** [comment]
I've created another pull request https://github.com/executors/issaquah_2016/pull/267 for the rest of the changes for query.

I've created a new type trait called `property_value` which is the type of the value that is returned from calling `query`. I've also improved on the requirements for the `query` customization point to require that the returned type, i.e. `property_value` be default constructible otherwise it's ill-formed and that it return a default constructed `property_value` if the function or member function are not well-formed.

We expect that any property which has a value associated with it would use `std::optional` or some other container type which allows the value to be optionally returned, however I've not enforced this in the interface. Instead I've put a note describing what the expected `property_value` would be for different properties.

I've also added a section to the explanatory document describing the `query` customization point and added some examples of using `std::optional`, `std::expected` and `std::status_value`.
